### ISSUE TITLE
Stop using pickle

### ISSUE THREAD

**Issue by QuLogic** [barsch/seishub.plugins.seismology#5]
I just noticed that SeisHub (and ObsPy) use pickle to communicate some things. Since this is a networked service, there are no guarantees that data may not be corrupt (accidentally or maliciously). There's even a great big red warning on the module documentation page for pickle about it.

**Comment by barsch** [comment]
the warning only states: "The pickle module is not intended to be secure against erroneous or maliciously constructed data. Never unpickle data received from an untrusted or unauthenticated source."

Personally I would never run SeisHub in an unsecured/untrusted network - one should use DMZ/Firewall/VPN etc. to ensure network security. So the only valid point here raised is the "erroneous or maliciously constructed data" - but the same may happen for JSON/XML or whatever format - a server has to handle corrupted data anyway in whatever format it comes ... so I fail to see whats the issue ...

**Comment by QuLogic** [comment]
Of course, as the author, you are intimately aware of the limitations. And anyone who takes a proper look at it would probably see that this is not supposed to be facing the Internet at large.

But the copy text for SeisHub really seems to indicate it's ready to go and while HTTPS or SSH are mentioned, they're not really _required_. At the very least, this is a documentation issue.

**Comment by barsch** [comment]
I just briefly checked the code to see where we use pickle (has been a while since I touched the code) - it seems for me that only the server sends pickled objects and does not retrieve/use them from clients - or did I miss something?

**Comment by QuLogic** [comment]
> So the only valid point here raised is the "erroneous or maliciously constructed data" - but the same may happen for JSON/XML or whatever format - a server has to handle corrupted data anyway in whatever format it comes ... so I fail to see whats the issue ...

Upon further reading, I think you may be underestimating the severity of that warning. For JSON or XML, the only problem is that the data may be corrupt. Assuming a perfectly correct parser, there is nothing that can go wrong.

With pickles, a perfectly correct parser allows _arbitrary_ code execution. An attacker can gain a remote shell or delete all your stuff or whatever they want (that your Python process can do, of course.)

> I just briefly checked the code to see where we use pickle (has been a while since I touched the code) - it seems for me that only the server sends pickled objects and does not retrieve/use them from clients - or did I miss something?

Yes, but I can't change only the client and not the server.

**Comment by barsch** [comment]
I knew that pickle should be avoided if possible but I wasn't aware that arbitrary code can be executed and in this context I agree too that it may be unwise to use pickle.

Just to make clear: SeisHub itself doesn't use pickle, while the seismology plugin uses it to create ObsPy stream objects on a clients request (e.g. obspy.seishub.client.getWaveform). While this never could compromise the server itself, it may be a security risk for a client. Again a question of trust. I personally know of only around 3 active SeisHub instances - so the number of potential attackers (read servers) are limited.

Three solutions I see:
(1) work on a way to serialize ObsPy stream objects into and from JSON in a reliable/secure fashion. Once this is done its quiet easy to switch to JSON on both server and clients. However pickle will stay for quiet a while in order to maintain backwards compatibility. Such serialization functionality would be handy for other projects too.

(2)  Completely rewrite getWaveform on client and server using a standard format to share fixed data -> we may loose metadata created by SeisHub which are not defined within this methods.

(3) Try SafeUnpickler class in http://michael-rushanan.blogspot.de/2012/10/why-python-pickle-is-insecure.html on client side

Other suggestions?

**Comment by barsch** [comment]
wontfix
